test/fake_http_server.h

```diff
@@ -31,15 +34,84 @@ class FakeServer {
   std::string GetUrl();
   void SetResponse(const std::string& path, const std::string& response);
 
+  // Indicates that for the given path, the server should stream
+  // responses over a hanging GET.
+  void AllowStream(const std::string& path);
+
+  // Blocks until at least one client has connected to the given path.
+  // Returns false if the timeout is reached with no client connections.
+  bool WaitForOneStreamWatcher(const std::string& path, time::seconds timeout);
+
+  // Sends a streaming response to all watchers for the given path.
+  void SendStreamResponse(const std::string& path, const std::string& response);
+
+  // Closes all open streams on the server.
+  void TerminateAllStreams();
+
  private:
   struct Handler;
   typedef boost::network::http::server<Handler> Server;
 
   // Handler that maps paths to response strings.
   struct Handler {
+    // Stream holds the state for an endpoint that streams data over a
+    // hanging GET.
+    //
+    // There is a queue for each client, and the mutex guards access
+    // to queues vector and any queue element.
+    //
+    // This struct only holds a pointer to each queue; it does not
+    // take ownership.
+    class Stream {
+     public:
+      void AddQueue(std::queue<std::string>* queue) {
+        {
+          std::lock_guard<std::mutex> lk(mutex_);
+          queues_.push_back(queue);
+        }
+        // Notify the condition variable to unblock any calls to
+        // WaitForOneStreamWatcher().
+        cv_.notify_all();
+      }
+
+      bool WaitForOneWatcher(time::seconds timeout) {
+        std::unique_lock<std::mutex> queues_lock(mutex_);
+        return cv_.wait_for(queues_lock,
+                            timeout,
+                            [this]{ return this->queues_.size() > 0; });
+      }
+
+      void SendToAllQueues(const std::string& response) {
+        {
+          std::lock_guard<std::mutex> lk(mutex_);
+          for (auto* queue : queues_) {
+            queue->push(response);
+          }
+        }
+        cv_.notify_all();
+      }
+
+      std::string GetNextResponse(std::queue<std::string>* queue) {
+        std::unique_lock<std::mutex> lk(mutex_);
+        cv_.wait(lk, [&queue]{ return queue->size() > 0; });
+        std::string s = queue->front();
```

Comment:
`const std::string&`

Reply:
Silly me. `queue->pop()` invalidates the reference...

Reply:
Acknowledged.